docs/proposals/remote-artifacts.md

```diff
@@ -0,0 +1,267 @@
+<!--
+Copyright The Shipwright Contributors
+
+SPDX-License-Identifier: Apache-2.0
+-->
+
+---
+title: remote-artifacts
+authors:
+  - "@otaviof"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-10-02
+last-updated:  2020-10-02
+status: provisional
+---
+
+Shipwright Remote Artifacts
+---------------------------
+
+# Summary
+
+Remote artifacts are often dependencies of the software building process, and hence are also
+required when dealing with container images. This enhancement proposal focuses on adding a build
+artifacts abstraction to Shipwright's Operator.
+
+# Motivation
+
+Give Shipwright's operator broader build use-case support by enhancing its capabilities to include
+the concept of Artifacts. In other words, remote entities that will be available for the image
+build process.
+
+End users will be allowed to declare artifacts that represent remote dependencies alongside
+Builds, and easily link them together. For example, a Java application which downloads certain
+jars into the classpath, a Node.js application adding external images, and so forth.
+
+## Goals
+
+*   Provide means to declare remote artifacts, dependencies that can be employed on builds;
+*   Create the mechanism to download and prepare remote artifacts for builds;
+
+## Non-Goals
+
+*   Automate the upload of local artifacts into the cluster;
+*   Manage remote artifacts;
+*   Amend container images created by the given `BuildStrategy`;
+*   Walking a remote directory tree, similarly to a `git checkout`;
+
+# Proposal
+
+The enhancement proposal is centered around the idea of declaring external dependencies, here
+called "artifacts", and being able to use them during the container image building process.
+
+## Build Artifacts
+
+The remote artifacts will be directly expressed in the body of a `Build` resource, as the
+following example:
+
+```yml
+---
+apiVersion: build.dev/v1alpha1
+kind: Build
+metadata:
+  name: nexus-jar
+spec:
+  artifacts:
+    - name: main-jar
+      url: https://nexus.company.com/app/main.jar
+      path: $(workspace)/classpath/main.jar
```

Comment:
Should we maybe consider extending the source section instead ?

```yaml
apiVersion: build.dev/v1alpha1
kind: Build
metadata:
  name: nexus-jar
spec:
  sources
    # A source of type=git must only exist once or be absent
    - type: git
      url: git@github.ibm.com:SCHWARZS/test-project.git
      revision: sascha-test
      credentials:
        name: ibm-github
    # A source of type=http can exist multiple times
    - type: http
      url: https://nexus.company.com/app/main.jar
      path: classpath/main.jar # Not sure if I would allow values that do not start with ${workspace} here but instead always have that as prefix
    # Credentials support for whatever we want, start with basic auth only, client-cert maybe in the future, we will need to determine it based on how the secret looks
    - type: http
      url: https://restricted.server.com/lib.jar
      credentials:
        name: restricted-server
      path: classpath/lib.jar
    # A source of type=http can have an optional extract flag set to true which means that path is interpreted as a directory instead of a file and the file content is extracted (TBD: can we autodetect the file type or do we need to make extract an enum or add another compressiontype parameter?)
    - type: http
      url: https://www.myserver.com/sources.zip
      extract: true
      path: sources
```

This would be an extensible mechanism. In the future we might support other types (loading from maybe some configmap where the cli placed the local sources).

Reply:
I think that would be hard. An HTTPS URL can be a public Git repository (especially as we do not require the `.git` suffix at the moment), a SubVersion repository, a plain file and probably more.

Reply:
I think given the distinction for types of data, one being VCS and the other remote artifacts, I think they should be described in different logical blocks. So, we also have a clear separation between those two concerts.

Reply:
@otaviof a VCS resource is also remote. There is "just" a different protocol to download them.

Reply:
> This would be an extensible mechanism. In the future we might support other types (loading from maybe some configmap where the cli placed the local sources).

`ConfigMap` is certainly a valid use case, though with a cli upload I see PersistentVolume/PVC being a more likely result.

Another thing we should clarify is that "sources" are things that are intended to be present in the resulting image. There is a separate use case for "volume" support (like caches), in which case the data present in the volume is not included in the resulting container image.

Reply:
> ```yaml
> ...
> spec:
>   sources
>     # A source of type=git must only exist once or be absent
>     - type: git
>       url: git@github.ibm.com:SCHWARZS/test-project.git
>       revision: sascha-test
>       credentials:
>         name: ibm-github

Wouldn't we want to employ the discriminator pattern?

```yaml
spec:
  sources:
  - name: git
    type: Git
    git:
      url: ...
...
```

(note that `type: Git` may not be required, apimachinery can fill it in).

Reply:
Same line of thought, I see a VCS as yet another remote artifact. I´m not able to understand the distinction.

Reply:
Please consider latest commit, I'm adding a initial `.spec.sources` support with you inputs.

Reply:
@otaviof thanks, I took a look.

> http: settings for http, namely path (optional);

Is the new field from above required? we already have something call `contextDir`, can this serve the same purpose?

@adambkaplan   @SaschaSchwarze0 @otaviof I have some concerns around `spec.sources`. First of all I think introducing a `type` is a nice way to distinguish the different assets(remote files, git,etc). I´m struggling to understand the value of moving from of a map(spec.source) into a list(spec.sources). My rational is:

- This proposal does not explains the use cases on why a list is needed
- This will break the current API
- `spec.source` is for me one of the structs that is very simple to understand, and I would like to keep it this way. Adding support for a list of sources is going to add complexity on that API level.
- supporting `spec.sources` will introduce some code complexity when generating the Taskrun. It would be good to  highlight those and decide if it worth the effort.

@sbose78 probably we want your feedback also. Also, do we have some good references on how Build v1 did it(shouldnt bias us)?, but it will help to understand the rational there.